Sources/TuistKit/Services/TestService.swift

```diff
@@ -371,35 +371,54 @@ final class TestService { // swiftlint:disable:this type_body_length
             graph: graph
         ) else { return }
 
-        for testScheme in testSchemes {
-            try await self.testScheme(
-                scheme: testScheme,
-                graphTraverser: graphTraverser,
-                clean: clean,
-                configuration: configuration,
-                version: version,
-                deviceName: deviceName,
-                platform: platform,
-                rosetta: rosetta,
-                resultBundlePath: resultBundlePath,
-                derivedDataPath: derivedDataPath,
-                retryCount: retryCount,
-                testTargets: testTargets,
-                skipTestTargets: skipTestTargets,
-                testPlanConfiguration: testPlanConfiguration,
-                passthroughXcodeBuildArguments: passthroughXcodeBuildArguments
-            )
-        }
-
         let uploadCacheStorage: CacheStoring
         if noUpload {
             uploadCacheStorage = try await cacheStorageFactory.cacheLocalStorage()
         } else {
             uploadCacheStorage = cacheStorage
         }
+
+        do {
+            for testScheme in testSchemes {
+                try await self.testScheme(
+                    scheme: testScheme,
+                    graphTraverser: graphTraverser,
+                    clean: clean,
+                    configuration: configuration,
+                    version: version,
+                    deviceName: deviceName,
+                    platform: platform,
+                    rosetta: rosetta,
+                    resultBundlePath: resultBundlePath,
+                    derivedDataPath: derivedDataPath,
+                    retryCount: retryCount,
+                    testTargets: testTargets,
+                    skipTestTargets: skipTestTargets,
+                    testPlanConfiguration: testPlanConfiguration,
+                    passthroughXcodeBuildArguments: passthroughXcodeBuildArguments
+                )
+            }
+        } catch {
+            // Check the test results and store successful test hashes for any targets that passed
+            guard let resultBundlePath, let resultParser = XCResultParser(url: resultBundlePath.url) else { throw error }
```

Comment:
```suggestion
            guard let resultBundlePath, let resultParser = xcresultParser.parse(resultBundlePath) else { throw error }
```

Let's:
- not use `URL` unless working directly with external dependencies. We should keep using `AbsolutePath` as much as possible
- let's pass the parser instance as a dependency, so we can more easily mock the value